snuba/datasets/cdc/cdcprocessors.py

```diff
@@ -1,26 +1,92 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Any, Mapping, Optional, Sequence, Type
+
 from snuba.processor import MessageProcessor
+from snuba.writer import WriterTableRow
 
 KAFKA_ONLY_PARTITION = 0  # CDC only works with single partition topics. So partition must be 0
 
 
+class CdcMessageRow(ABC):
+    """
+    Takes care of the data transformation from WAL to clickhouse and from
+    bulk load to clickhouse. The goal is to keep all these transformation
+    function in the same place because they ultimately have to be consistent
+    with the Clickhouse schema.
+    """
+
+    @classmethod
+    def from_wal(
+        cls,
+        offset: int,
+        columnnames: Sequence[str],
+        columnvalues: Sequence[Any],
+    ) -> CdcMessageRow:
+        raise NotImplementedError
+
+    @classmethod
+    def from_bulk(
+        cls,
+        row: Mapping[str, Any],
+    ) -> CdcMessageRow:
+        raise NotImplementedError
+
+    @abstractmethod
+    def to_clickhouse(self) -> WriterTableRow:
+        raise NotImplementedError
+
+
 class CdcProcessor(MessageProcessor):
 
-    def __init__(self, pg_table):
+    def __init__(self, pg_table: str, message_row_class: Type[CdcMessageRow]):
         self.pg_table = pg_table
+        self._message_row_class = message_row_class
 
-    def _process_begin(self, offset):
+    def _process_begin(self, offset: int):
         pass
 
-    def _process_commit(self, offset):
+    def _process_commit(self, offset: int):
         pass
 
-    def _process_insert(self, offset, columnnames, columnvalues):
-        pass
+    def _process_insert(self,
+        offset: int,
+        columnnames: Sequence[str],
+        columnvalues: Sequence[Any],
+    ) -> Optional[WriterTableRow]:
```

Comment:
This is helpful, I didn't realize these are optional. Why does this differ than return type of `CdcMessageRow.to_clickhouse()`? Shouldn't they be equivalent?

Reply:
They differ because I wanted to let the dataset specific class decide whether to ignore a WAL row. For example we could have a dataset without deletions. In that case WAL deletion should produce no effect and just return None. Essentially we let the dataset a bit more autonomy on how to interpret the WAL if they want to override these methods.
On the other hand, when we have a valid CdcMessageRow instantiated, that has to be serialized into a write to clickhouse. I cannot think about a legit case where we would be able to instantiate CdcMessageRow but we may decide to return None